### PR TITLE
Accessing layer and stereo from rawId

### PR DESCRIPTION
**Pull request by alesaggio**

### CODE REVIEW

File: CalibTracker/SiStripLorentzAngle/interface/LA_Filler_Fitter.h [hunk 1]
```diff
@@ -4,8 +4,7 @@
 #include <string>
 #include <vector>
 #include <map>
-#include "DataFormats/SiStripDetId/interface/TOBDetId.h"
-#include "DataFormats/SiStripDetId/interface/TIBDetId.h"
+#include "DataFormats/SiStripDetId/interface/SiStripDetId.h"
```
Comment (OlivierBondu) [inline CalibTracker/SiStripLorentzAngle/interface/LA_Filler_Fitter.h:7]: This shouldn't be necessary if the include is already defined in the `TIBorTOBparameters.h` ?

**Review by OlivierBondu — CHANGES_REQUESTED** [review]
Looks good to me! only one minor thing that I can see, I guess the `TIBorTOBparameters.h` cannot be replaced by something already existing in central space ?

File: CalibTracker/SiStripLorentzAngle/interface/LA_Filler_Fitter.h (hunk 1 above)
Comment (pieterdavid) [reply CalibTracker/SiStripLorentzAngle/interface/LA_Filler_Fitter.h:7]: indeed, but it shouldn't be included in `TIBorTOBparameters.h` in the first place, IIRC (not sure if it needs to be included here then)

File: CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#ifndef TIBORTOBPARAMETERS_H
+#define TIBORTOBPARAMETERS_H
+
+#include "DataFormats/SiStripDetId/interface/SiStripDetId.h"
```
Comment (pieterdavid) [inline CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h]: you don't need this one (include `DataFormats/DetId/interface/DetId` instead)

File: CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+
+uint32_t getTIBOrTOBLayer( DetId detId )
+{
+  return ((detId.rawId()>>14) & 0x7);
+};
+
+bool getTIBorTOBStereo( DetId detId )
+{
+  return ( ((detId.rawId()>>0) & 0x3) == 1 ) ? 1 : 0;
```
Comment (pieterdavid) [inline CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h]: `return (detId.rawId() & 0x3) == 1;` should be sufficient

File: CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#ifndef TIBORTOBPARAMETERS_H
+#define TIBORTOBPARAMETERS_H
+
+#include "DataFormats/SiStripDetId/interface/SiStripDetId.h"
+
+// WARNING: This header file is created on the assumption
+// that the subDetector Id (TIB or TOB) is already known.
+// It is used inside CalibTracker/SiStripLorentzAngle only.
```
Comment (pieterdavid) [inline CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h]: How about something like "WARNING: these methods assume that the DetId is a TIBDetId or TOBDetId (and don't check this). Please use the corresponding TrackerTopology methods whenever possible.
Code copied from TIBDetId and TOBDetId."

File: CalibTracker/SiStripLorentzAngle/src/LA_Filler.cc [hunk 5]
```diff
@@ -85,11 +86,11 @@ granularity(const SiStripDetId detid, const float tthetaL, const Long64_t TFE_in
...
 
 std::string LA_Filler_Fitter::
-subdetLabel(const SiStripDetId detid) { return detid.subDetector()==SiStripDetId::TOB? "TOB" : "TIB";}
+subdetLabel(const SiStripDetId detid) { return detid.subDetector()==SiStripDetId::TOB ? "TOB" : "TIB";}
 std::string LA_Filler_Fitter::
 moduleLabel(const SiStripDetId detid) { return subdetLabel(detid) + "_module"+boost::lexical_cast<std::string>(detid());}
 std::string LA_Filler_Fitter::
 layerLabel(const SiStripDetId detid) {
-  unsigned layer = detid.subDetector() == SiStripDetId::TOB ? TOBDetId(detid()).layer() : TIBDetId(detid()).layer();
+  unsigned layer = detid.subDetector() == getTIBOrTOBLayer(detid);
```
Comment (pieterdavid) [inline CalibTracker/SiStripLorentzAngle/src/LA_Filler.cc]: should be `unsigned layer = getTIBOrTOBLayer(detid);` (the current code will compare layer do subdetector numbers)

**Review by pieterdavid — CHANGES_REQUESTED** [review]
A few issues left (one bug, the others minor)

**Comment by pieterdavid** [comment]
@OlivierBondu wrote:
> Looks good to me! only one minor thing that I can see, I guess the TIBorTOBparameters.h cannot be replaced by something already existing in central space ?

Yes, that's what `TIBDetId` and `TOBDetId` were. Since this is essentially a ROOT macro in disguise (all the work is done from the constructor of the `ESProducer`), we don't have a way to access the `edm::EventSetup` without rewriting the whole thing, that's why we'd suggest to add the specific pieces of code here (in such a way that it's not easy to use from elsewhere ;-) ). Actually everything could be moved into the `plugins` directory, nothing depends on `CalibTracker/SiStripLorentzAngle`.
Opinions/thoughts welcome ;-)

File: CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h (hunk 4 above)
Comment (OlivierBondu) [reply CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h]: +1, maybe also specify in which package/subpackage the TIBDetId are, we're going to deprecate them so you want to ease life of people possibly digging history some day...

**Review by OlivierBondu — APPROVED** [review]

File: CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h (hunk 4 above)
Comment (pieterdavid) [reply CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h]: maybe "see `git show CMSSW_9_2_3_patch2:DataFormats/SiStripDetId/interface/TIBDetId.h` and `git show CMSSW_9_2_3_patch2:DataFormats/SiStripDetId/interface/TOBDetId.h`", or is that too detailed?

**Review by pieterdavid — APPROVED** [review]
Looks good now, thanks! As mentioned before: we'll have to explain a bit in the PR why we would go for this solution ;)

File: CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h (hunk 4 above)
Comment (OlivierBondu) [reply CalibTracker/SiStripLorentzAngle/interface/TIBorTOBparameters.h]: that sounds perfect :)

**Comment by OlivierBondu** [comment]
this can be closed if it is outdated by #10 ?

**Pull request closed without merging**